### PR TITLE
Add priority loading for model entities

### PR DESCRIPTION
**Pull request by huffman**
Putting this up for Chris. No need to review.

### CODE REVIEW

File: libraries/entities-renderer/src/EntityTreeRenderer.h [hunk 1]
```diff
@@ -46,6 +46,8 @@ class EntityTreeRenderer : public OctreeRenderer, public EntityItemFBXService, p
     virtual PacketType getExpectedPacketType() const { return PacketType::EntityData; }
     virtual void setTree(OctreePointer newTree);
 
+    glm::vec3 cameraPosition {};
```
Comment (ZappoMan) [inline libraries/entities-renderer/src/EntityTreeRenderer.h]: I am not a fan of this... the camera is independent of the tree... there may often be multiple cameras looking at the same tree.

Comment (huffman) [inline libraries/entities-renderer/src/EntityTreeRenderer.h]: I'm just trying to get a build to test angular loading in the easiest possible way. As noted in the PR title and description, there is no need to review this PR.

That said, that piece of code isn't actually used and was dragged in unintentionally.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4aa57c7aa8b58146451b74119ce07090eb15866d available at: https://gist.github.com/1463194eb9cd52be934bf0c2897d329e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-4aa57c7.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-4aa57c7.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 61d7416463139c1e72ce4e8958a9d9c2dbcc5ace available at: https://gist.github.com/1463194eb9cd52be934bf0c2897d329e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-61d7416.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-61d7416.dmg

**Comment by ZappoMan** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: d9e74c5957835b552b8cb4bc8507ef6c7a757ce4 available at: https://gist.github.com/1463194eb9cd52be934bf0c2897d329e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-d9e74c5.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-d9e74c5.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5316ee011f144879c7f9cb4a849c545de8213d21 available at: https://gist.github.com/1463194eb9cd52be934bf0c2897d329e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-5316ee0.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-5316ee0.dmg

**Comment by huffman** [comment]
The goal of this change is to improve domain loading by prioritizing the order in which entities are loaded by their angular size. For identically sized items this will mean that you will prioritize the items closest to you to be loaded first.

Because we download assets for multiple entities in parallel, and because not all entity assets take the same time to download, it's possible for entities with lower priority to have their assets appear first.

To test:

1. Go to your home domain at 0,0,0
2. Run https://gist.githubusercontent.com/cozza13/12a5ca2e5ad01b1a1224cd306d26d163/raw/fd0a1d2c4022201efe25da896ccf39e68463b032/createmuggrid.js to create a grid of cups
3. Go to one corner of the cups and position yourelf so that you can see all of the cups
4. Go to another domain of your choosing
5. Click Edit > Reload Content
6. Open the address bar and press the back button
7. Confirm that the mugs closest to you load first
8. Repeat steps 3-7 for the opposite corner

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9a468e3738b40c37e5c69ae969ade9e860ba0713 available at: https://gist.github.com/1463194eb9cd52be934bf0c2897d329e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-9a468e3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-9a468e3.dmg

**Comment by huffman** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9a468e3738b40c37e5c69ae969ade9e860ba0713 available at: https://gist.github.com/1463194eb9cd52be934bf0c2897d329e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-9a468e3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-9a468e3.dmg

**Comment by huffman** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9a468e3738b40c37e5c69ae969ade9e860ba0713 available at: https://gist.github.com/659e4902844db8874452de1226aa80c7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-9a468e3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-9a468e3.dmg

**Comment by huffman** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9a468e3738b40c37e5c69ae969ade9e860ba0713 available at: https://gist.github.com/1463194eb9cd52be934bf0c2897d329e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-9a468e3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-9a468e3.dmg

**Comment by Nex-Pro** [comment]
Result: **FAIL**

Followed steps above.

Also tested from other two corners of grid. ( marked nearest model position with a cube for reference )

Results were patchy at best. Although overall there was a tendency towards the nearer Items appearing first, on many occasions this was not the  case.

***In each of these images the nearest mug should be the lowest in the image.***

pr8311issue
pr8311issue2
pr8311issue3
pr8311issue4
pr8311issue5
pr8311issue6

**Comment by huffman** [comment]
@Nex-Pro You are right, it looks random in the PR build. Looking into it.

**Comment by huffman** [comment]
@Nex-Pro  Ok, I tested more. It's working, but with only 50 cups with relatively small file sizes the effect is largely hidden.

There are two ways to test this more easily, in order of difficulty:

1. Use a script that creates a grid of 10x10 mugs. Once the initial mugs load you should see that they generally load closer to you. https://gist.githubusercontent.com/huffman/8c20b4065a8c2ac6f19e234cd137c3d9/raw/6fe7f05504fc8beaec593e421df26536aa479d2a/grid.js
2. The reason this PR is difficult to test is that we have up to 16 downloads going at once, along with entities that come in at different times. This PR adds a parameter that can be passed to interface to tell it to cap the number of concurrent downloads to a different number - in this case, capping it to 1 makes it easier to see what is happening. To do this: go to interface.exe (the default directory for this PR would be `C:\Program Files\High Fidelity - PR8311`), right click, click `Create Shortcut`. Right click the shortcut, click `Properties`, then modify the `Target` by appending ` --concurrent-downloads 1`. It should look something like this: `"C:\Program Files\High Fidelity - PR8311\interface.exe" --concurrent-downloads 1`. Click the shortcut to run interface with only 1 concurrent download (you should be able to confirm this by clicking on the stats to see current downloads).

**Comment by Nex-Pro** [comment]
@huffman The mugs we are using should appear to the system as individual enties as I scripted the generation of them in Blender. ( Used a mug I created )  They all have individual mesh names file names and folders. Thoys then scipted the grid generator.

**Comment by Nex-Pro** [comment]
@huffman I'm happy to do tests 1 and 2 if you require that.

**Comment by huffman** [comment]
@Nex-Pro Ah, gotcha, I didn't see that! Chris had modified it for testing something different that required 50 or fewer entities, I should have used the original.

Yes, if you could try #2 to set the number of concurrent downloads that would be great. Thanks!

**Comment by Nex-Pro** [comment]
@huffman  That script rezzes all the mugs in the same spot.

**Comment by Nex-Pro** [comment]
@huffman  Correction the PR is making the objects all rez at center ( 0 , 0, 0 )

**Comment by huffman** [comment]
@Nex-Pro Does that happen when you refresh the cache? I've noticed that happening, @cozza13 mentioned that he's seen the same on other PRs. If you leave and come back it should correct itself.

**Comment by Nex-Pro** [comment]
@huffman , apologies. I forgot to continue with this.

**Comment by huffman** [comment]
Test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9a468e3738b40c37e5c69ae969ade9e860ba0713 available at: https://gist.github.com/1463194eb9cd52be934bf0c2897d329e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-9a468e3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-9a468e3.dmg

**Comment by Nex-Pro** [comment]
@huffman  Do you need the most recent build of this to be tested ?

**Comment by huffman** [comment]
@Nex-Pro I pushed an update that includes the latest, once that builds it will be ready for testing. Thanks!

**Comment by hifi-gustavo** [comment]
Build console output for commit: d367583426e0728a3d89bb6b8e83e678b197c156 available at: https://gist.github.com/1463194eb9cd52be934bf0c2897d329e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-d367583.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-d367583.dmg

**Comment by Nex-Pro** [comment]
@huffman  the console output for windows looks ok, should I just grab it now ?

**Comment by huffman** [comment]
test this please

**Comment by huffman** [comment]
@Nex-Pro Yeah, go for it. The build error was not related to the code.

**Comment by hifi-gustavo** [comment]
Build console output for commit: d367583426e0728a3d89bb6b8e83e678b197c156 available at: https://gist.github.com/1463194eb9cd52be934bf0c2897d329e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-d367583.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-d367583.dmg

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -3240,6 +3247,13 @@ void Application::init() {
         getEntities()->setViewFrustum(_viewFrustum);
     }
 
+    getEntities()->setEntityLoadingPriorityFunction([this](const EntityItem& item) {
+        auto dims = item.getDimensions();
+        auto maxSize = glm::max(dims.x, dims.y, dims.z);
+        auto distance = glm::distance(getMyAvatar()->getPosition(), item.getPosition());
+        return atan2(maxSize, distance);
```
Comment (sethalves) [inline interface/src/Application.cpp]: is this defined if distance is zero?

File: libraries/entities-renderer/src/EntityTreeRenderer.h [hunk 3]
```diff
@@ -66,7 +73,7 @@ class EntityTreeRenderer : public OctreeRenderer, public EntityItemFBXService, p
     void reloadEntityScripts();
 
     /// if a renderable entity item needs a model, we will allocate it for them
-    Q_INVOKABLE ModelPointer allocateModel(const QString& url, const QString& collisionUrl);
+    Q_INVOKABLE ModelPointer allocateModel(const QString& url, const QString& collisionUrl, float loadingPriority = 0);
```
Comment (sethalves) [inline libraries/entities-renderer/src/EntityTreeRenderer.h]: 0.0f ?

File: libraries/render-utils/src/Model.h [hunk 4]
```diff
@@ -405,6 +407,10 @@ public slots:
 
     bool _visualGeometryRequestFailed { false };
     bool _collisionGeometryRequestFailed { false };
+
+private:
+    float _loadingPriority { 0 };
```
Comment (sethalves) [inline libraries/render-utils/src/Model.h]: 0.0f

File: interface/src/Application.cpp (hunk 2 above)
Comment (huffman) [inline interface/src/Application.cpp]: Yes, but not if maxSize is `0` - I'm pretty sure we don't allow a dimension of 0, but I can make this explicit

File: interface/src/Application.cpp [hunk 5]
```diff
@@ -3240,6 +3247,13 @@ void Application::init() {
         getEntities()->setViewFrustum(_viewFrustum);
     }
 
+    getEntities()->setEntityLoadingPriorityFunction([this](const EntityItem& item) {
+        auto dims = item.getDimensions();
```
Comment (sethalves) [inline interface/src/Application.cpp]: I can't quite get my head around the threading here.  Does the EntityTree need to be locked here?

File: interface/src/Application.cpp (hunk 2 above)
Comment (sethalves) [inline interface/src/Application.cpp]: nm, man-page says it is

File: interface/src/Application.cpp (hunk 5 above)
Comment (huffman) [inline interface/src/Application.cpp]: This gets called from `RenderableModelEntityItem::getModel` which does make updates to itself (the entity item), so my expectation would be that it is safe. @ZappoMan Any thoughts on this?

File: interface/src/Application.cpp (hunk 2 above)
Comment (sethalves) [inline interface/src/Application.cpp]: If y is +0 (-0) and x is less than 0, +pi (-pi) is returned.

       If y is +0 (-0) and x is greater than 0, +0 (-0) is returned.

I think you're ok

**Comment by Nex-Pro** [comment]
@huffman  This has a tendency towards rezzing nearest first but reliability is less than good. Also I am seeing a strange thing.  I create a large cube ( as a reference point to fly towards with Y rot at 45 deg.  I place the cube at one of the corner mugs. 
Clearing the cache has the effect of moving the reference cube to 0,0,0

If I reload content and fly to the opposite corner of the grid the rez order is not updated so I can get there and watch the mugs rez towards me.

**Comment by Nex-Pro** [comment]
I tested this multiple times from each of the four corners of the grid.

**Comment by huffman** [comment]
Interesting. Did you try while setting the number of concurrent downloads to 1?

> but reliability is less than good. 
Reliability in loading the nearest first or has the stability of the application been affected?

**Comment by Nex-Pro** [comment]
@huffman 
Ah, I had a shortcut with concurrent at 1
 Hang on.

**Comment by huffman** [comment]
@Nex-Pro When using `reload content`, I would recommend going to another domain first, reloading the content, and then going back.

**Comment by hifi-gustavo** [comment]
Build console output for commit: fb14628a7ebff15c9bbf70d839aa83730f1d624e available at: https://gist.github.com/1463194eb9cd52be934bf0c2897d329e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-fb14628.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8311-fb14628.dmg

**Comment by Nex-Pro** [comment]
/me sets up a tent and waits for content to reload

**Comment by Nex-Pro** [comment]
@huffman  With concurrent downloads set to 1 this seems to work well albeit extremely slowly.

And as I mentioned before once the rezzing has begun the order is effectively carved in stone.

**Comment by huffman** [comment]
@Nex-Pro Great. The priority of an entity is determined when it first appears, so that is as expected. A future update may change that.

**Comment by huffman** [comment]
Thanks for testing @Nex-Pro

File: interface/src/Application.cpp (hunk 5 above)
Comment (ZappoMan) [inline interface/src/Application.cpp]: I think this is reasonably safe.

**Comment by Nex-Pro** [comment]
@huffman YVW

**Pull request merged**